### PR TITLE
Lifting function types

### PR DESCRIPTION
**Pull request by kvark**
Based on #71

### CODE REVIEW

**Comment by kvark** [comment]
@antiagainst FYI, CI is green on all three commits here. So if the code is good, we can merge the PR without squashing and avoid some churn

File: autogen/sr.rs [hunk 1]
```diff
@@ -243,6 +248,12 @@ pub fn gen_sr_code_from_instruction_grammar(
         // Get the token for its enumerant
         let inst_name = &inst.opname[2..];
         let name_ident = Ident::new(inst_name, Span::call_site());
+        let type_name = if inst.opname.len() > 6 {
```
Comment (antiagainst) [inline autogen/sr.rs]: What is this mysterious magic number here? Could you add some comment to explain? Thanks!

File: rspirv/sr/autogen_ops.rs [hunk 2]
```diff
@@ -18,9 +18,31 @@
 
 use crate::sr::{Token, Type};
 #[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Branch {
+    Branch {
+        target_label: spirv::Word,
```
Comment (antiagainst) [inline rspirv/sr/autogen_ops.rs]: So `spirv:Word` is the DR's way to "link" instructions. Here I think target_label should be `Token<BasicBlock>`.

File: rspirv/sr/autogen_ops.rs [hunk 3]
```diff
@@ -18,9 +18,31 @@
...
+    },
+    BranchConditional {
+        condition: spirv::Word,
```
Comment (antiagainst) [inline rspirv/sr/autogen_ops.rs]: Here `condition` should be `Token<Instruction>` and labels should be `Token<BasicBlock>`.

If you intend to change it later, would be nice at least to have some TODO in the source code. The existing implementation seems wrong to me.

File: rspirv/sr/autogen_ops.rs [hunk 4]
```diff
@@ -18,9 +18,31 @@
...
+        true_label: spirv::Word,
+        false_label: spirv::Word,
+        branch_weights: Vec<u32>,
+    },
+    Switch {
+        selector: spirv::Word,
+        default: spirv::Word,
+        target: Vec<(u32, Token<Type>)>,
```
Comment (antiagainst) [inline rspirv/sr/autogen_ops.rs]: Why `Token<Type>`? Shouldn't it be `Token<BasicBlock>`?

File: rspirv/sr/autogen_ops.rs [hunk 5]
```diff
@@ -18,9 +18,31 @@
...
+    },
+    Return,
+    ReturnValue {
+        value: spirv::Word,
```
Comment (antiagainst) [inline rspirv/sr/autogen_ops.rs]: Token<Instruction>

**Comment by kvark** [comment]
@antiagainst this is rebased and reduced in scope, please take a look.

**Comment by antiagainst** [comment]
Smaller ones are much easier to review! :)

**Pull request merged**